flink-core/src/main/java/org/apache/flink/configuration/SecurityOptions.java

```diff
@@ -217,6 +217,16 @@ public class SecurityOptions {
                     .withDescription(
                             "Turns on SSL for external communication via the REST endpoints.");
 
+    @Documentation.Section(Documentation.Sections.EXPERT_SECURITY_SSL)
+    public static final ConfigOption<String> SSL_REST_SSL_CONTEXT_SUPPLIER =
+            key("security.ssl.rest.ssl-context-supplier")
+                    .stringType()
+                    .noDefaultValue()
+                    .withDescription(
+                            "A fully qualified class name that implements the Supplier<SslContext> interface."
+                                    + " The implementation must have a public constructor with the signature"
+                                    + " (Configuration config, boolean isClient, SslProvider sslProvider)");
```

Comment:
How about `interface SslContextSupplier implements Supplier<SslContext>`?

The other option that I was considering was doing something like:
```
interface SslContextSupplierFactory {
   Supplier<SslContext> create(Configuration config, boolean isClient, SslProvider sslProvider)
}
```

which then also removes the need for a magic constructor, but I wasn't a huge fan of the "SupplierFactory" java-ism.  Thoughts?

Reply:
So now I see, you want to spare an if condition like `if (reloadableSslContext == null) { // create } else { // give back }`.
I would vote one not doing a factory of factory:
```
interface SslContextSupplier {
   SslContext get(Configuration config, boolean isClient, SslProvider sslProvider)
}
```
Get can decide whether new or already existing instance needed. Feel free to change the naming but:
* putting API into doc
* factory of a factory 

is not something we would like to do.

Reply:
I'd prefer to keep the `Supplier<SslContext>` interface at the end (or something similar) since it makes passing it around much simpler.  In your interface for example you'll need the Configuration, clientMode, and SslProvider at the time when you need to get the SslContext.

For example, if the interface looked like this, `SSLHandlerFactory` would also need to then have access to the configuration and SslProvider.

Reply:
Easy to pass inside the code should never drive the API. Having a factory of a factory is -1 from my side.

Reply:
It's not just easy to pass, its keeping good abstraction between components.  The consumers of SslContexts shouldn't need to know how to create them imo, merging both the factory and supplier couples those concerns together.

Reply:
ok, updated.  fwiw I really don't like the way the interface works, I think it makes it really clunky for implementations since they'll need to make the get method lazy, and handle cases where the parameters to get could change between calls, but if you're happy with it that's good enough I guess.